instrumentctl/g9_driver.py

```diff
@@ -0,0 +1,255 @@
+    5: 'Failure of the associated dual-channel input'
+}
+
+outStatus = {
+    0: 'No error',
+    1: 'Invalid configuration',
+    2: 'Overcurrent detection',
+    3: 'Short circuit detection',
+    4: 'Stuck-at-high detection',
+    5: 'Failure of the associated dual-channel output',
+    6: 'Internal circuit error',
+    8: 'Dual channel violation'
+}
+
+usStatus = {
+    9: "Output Power Supply Error Flag",
+    10: "Safety I/O Terminal Error Flag",
+    13: "Function Block Error Flag"
+}
+
+NUMIN = 13
+
+SNDHEADER = b'\x40\x00\x00\x0F\x4B\x03\x4D\x00\x01' 
+SNDDATA = b'\x00\x00\x00\x00'
+SNDRES = b'\x00\x00'
+
+RECHEADER = b'\x40\x00\x00'
+
+FOOTER = b'\x2A\x0D'
+
+class G9Driver:
+    def __init__(self, port=None, baudrate=9600, timeout=0.5, logger=None, debug_mode=False):
+        # not sure if we will need all of these but just adding them now in case
+        self.US, self.SITDF, self.SITSF, self.SOTDF, self.SOTSF, self.lastResponse = None, None, None, None, None, None
+        if port:
+            self.ser = serial.Serial(port, baudrate, parity=serial.PARITY_EVEN,
+                                      stopbits=serial.STOPBITS_ONE, bytesize=serial.EIGHTBITS,
+                                        timeout=timeout)
+        self.debug_mode = debug_mode
+        self.logger = logger
+        self.NUMIN = NUMIN
+        self.input_flags = []
+
+    def send_command(self):
+        if not self.is_connected():
+            raise ConnectionError("Seiral Port is Not Open.")
+
+        checksum = self.calculate_checksum(SNDHEADER + SNDDATA + SNDRES, 0 , len(SNDHEADER + SNDDATA))
+
+        self.ser.write(SNDHEADER + SNDDATA + SNDRES + checksum + FOOTER)
+
+        self.response()
+
+    # used mainly for the check sum but can also be used to check for error flags
+    # needs an input of a byte string and the range of bytes that need to be sum
+    # will return the sum of the bytes in the a byte string in the form of b'\x12'
+    def calculate_checksum(self, byteString, startByte, endByte):
+        assert isinstance(byteString, bytes)
+        return sum(byteString[startByte:endByte + 1]).to_bytes(2, "big") 
+
+    # helper function to convert bytes to bits for checking flags
+    # not currently being used but many be helpful in the future for getting errors
+    def bytes_to_binary(self, byte_string):
+        return ''.join(format(byte, '08b') for byte in byte_string)
+    
+    # this method is made to check the error flags, right not only checks the last 13 bits
+    # of a byte string
+    def check_flags13(self, byteString, norm = '1'):
+        assert isinstance(byteString, bytes)
+        binary_string = self.bytes_to_binary(byteString)[-NUMIN:]
+        string_of_ones = norm * NUMIN
+        return binary_string == string_of_ones
+    
+    def response(self):
+        if not self.is_connected():
+            raise ConnectionError("Serial Port is Not Open.")
+        
+        data = self.ser.read_until(b'\r')
```

Comment:
This still doesn't follow the G9 manual:
![image](https://github.com/user-attachments/assets/0aa62c90-4a22-4e84-a93e-7e1686ef5869)

I can't think of a good reason not to read until the full footer is received. should be:
```
data = self.ser.read_until(FOOTER)
```